src/storage/core.rs

```diff
@@ -0,0 +1,471 @@
+use futures::{
+    executor::block_on,
+    future::{FutureExt, FutureObj},
+    lock::Mutex,
+    stream::{futures_unordered, Stream, StreamExt},
+    task::{self, Poll, Spawn, SpawnExt, Waker},
+};
+use std::{
+    fs::{self, DirEntry, File, OpenOptions},
+    io,
+    path::{Path, PathBuf},
+    pin::Pin,
+    sync::{
+        atomic::{AtomicBool, AtomicUsize, Ordering},
+        Arc,
+    },
+    thread,
+    time::{Duration, Instant},
+};
+
+use crate::{
+    blob::{self, Blob},
+    record::Record,
+};
+
+const BLOB_FILE_EXTENSION: &str = "blob";
+const LOCK_FILE: &str = "pearl.lock";
+
+/// # Description
+/// A specialized Result type
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// # Description
+/// Used to create a storage, configure it and manage
+/// # Examples
+/// ```
+/// use pearl::{Storage, Builder};
+///
+/// let builder = Builder::new()
+///     .work_dir("/tmp/pearl/")
+///     .max_blob_size(1_000_000)
+///     .max_data_in_blob(1_000_000_000)
+///     .blob_file_name_prefix("enough");
+/// ```
+/// let mut storage = builder.build().unwrap();
+/// storage.init().unwrap();
+/// ```
+#[derive(Debug)]
+pub struct Storage {
+    twins_count: Arc<AtomicUsize>,
+    shared: Arc<Shared>,
+    inner: Arc<Mutex<Inner>>,
+}
+
+impl Drop for Storage {
+    fn drop(&mut self) {
+        let twins = self.twins_count.fetch_sub(1, Ordering::Relaxed);
+        // 1 is because twin#0 - in observer tread, twin#1 - self
+        if twins <= 1 {
+            self.shared.observer_state.store(false, Ordering::Relaxed);
```

Comment:
Can we give more straight forward name for `observer_state` something like `need_exit`